ADR/0056-schema-goverened-correlation-id.md

```diff
@@ -0,0 +1,57 @@
+# ADR-0056: Schema-Governed Correlation ID for Workflow Traceability
+
+Date 2023-11-17
+
+## Status
+
+Proposed
+
+## Context
+
+Konflux CI orchestrates multi-stage workflows involving Build PipelineRuns for individual `Component` resources, followed by creation and reconciliation of `Snapshot` resources, execution of `IntegrationTestScenario` PipelineRuns, and finally Release CRs and Release PipelineRuns. Although these steps represent a single logical workflow, Konflux currently lacks a deterministic way to correlate all PipelineRuns generated by the same webhook event or logical change.
+
+Tekton’s built-in identifiers (such as event IDs) are random and cannot be reproduced. CEL templates can extract event metadata but cannot compute stable hashes. PipelineRuns may be short-lived or created in remote clusters, preventing reliable correlation based on their presence in the cluster. Downstream observability systems that consume PipelineRun CloudEvents require a durable, deterministic identifier spanning the full Build -> Integration -> Release lifecycle.
+
+A correlation mechanism is needed that is deterministic across retries and replays, configurable on a per-namespace basis, SCM-agnostic, and capable of expressing both single-event and multi-event fan-in, all without modifying existing CRDs.
+
+## Decision / Proposal
+
+Konflux introduces a **schema-governed correlation identifier** carried through the system using a single label:
+
+```
+pipelinesascode.tekton.dev/correlation-id
+```
+
+Two identifiers are defined:
+
+**Event Correlation ID (ECID).**
+Generated by Pipelines-as-Code (PaC) per webhook event. Each namespace may define a `konflux-correlation-policy` Secret containing a UUIDv5 namespace, a selected set of normalized SCM fields (provider, repository, commit, PR number, event ID, etc.), and a canonical JSON template. PaC watches and caches this policy; for each webhook event, it normalizes the selected fields, renders the template, computes a UUIDv5 hash, and labels all PipelineRuns created for that event with the resulting ECID.
+
+**Snapshot Correlation ID (SCID).**
+Generated by Integration when a Snapshot contains components produced from multiple ECIDs. Build-service records the ECID in each Component’s build metadata when a Build PipelineRun completes. When Integration creates a Snapshot, it copies these ECIDs into the Snapshot’s component list. On reconciliation, if all components share the same ECID, that value becomes the Snapshot’s correlation ID; if multiple ECIDs are present, Integration computes a UUIDv5 over the sorted distinct ECIDs and uses it as the SCID. Parent ECIDs are stored in paged Snapshot annotations along with page count, total count, and a fingerprint.
```

Comment:
> Build-service records the ECID in each Component’s build metadata when a Build PipelineRun completes.

What build metadata are you referring to? The build service doesn't propagate anything to resources on the cluster. It is actually the integration service which updates the Component CR with the latest known image build, but this is only so that it knows what references to use for the Snapshot generation.